torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -445,47 +448,46 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::executeSendFunctionAsync(
         /*incrementOutstandingTasks=*/false);
 
     // Build the 'uber' future that waits for everything.
-    auto callbackFuture = std::make_shared<rpc::FutureMessage>();
+    auto callbackFuture =
+        std::make_shared<c10::ivalue::Future>(c10::NoneType::create());
+
+    accumulateGradFuture->addCallback([autogradContext,
+                                       callbackFuture,
+                                       accumulateGradFuture]() {
```

Comment:
IIUC, this circular reference won't leak, because future swaps callbacks before executing them. Do we need to add a comment in ivalue Future to highlight this requirement?

Reply:
Isn't this kind of the convention we use with ivalue::Future today to get the value of the future in the callback? We do this in many places today (ex: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/request_callback_impl.cpp#L168)